### PR TITLE
FEATURE: content based filtering 

### PR DESCRIPTION
**Pull request by nselpriv**
### This Pull request adds the following functionality

1. New articles like function in elastic-recommender that uses user likes to recommend articles like what they have already explicitly liked 
2. New endpoint for /foryou  

**This PR is combined with a PR in zeeguu-web, they should be approved together https://github.com/zeeguu/web/pull/371**

### CODE REVIEW

File: zeeguu/core/content_recommender/elastic_recommender.py [hunk 1]
```diff
@@ -303,3 +302,78 @@ def _difficuty_level_bounds(level):
...
+    es = Elasticsearch(ES_CONN_STRING)
+    fields = ["content", "title"]
+    language = Language.find_by_id(language_id)
+    like_documents = [
+        {"_index": ES_ZINDEX, "_id": str(doc_id) } for doc_id in recommended_articles_ids
+    ]
+
+    cutoff_date = datetime.now() - timedelta(days=article_age)
+
+    mlt_query = {
```
Comment (tfnribeiro) [inline zeeguu/core/content_recommender/elastic_recommender.py:317]: I would suggest building this query by using a pattern similar of what is used in `build_elastic_recommender_query` under `elastic_query_builder.py`.

Comment (nselpriv) [reply zeeguu/core/content_recommender/elastic_recommender.py]: @tfnribeiro 
Implemented your suggestion 👍

File: zeeguu/core/elastic/elastic_query_builder.py [hunk 2]
```diff
@@ -182,3 +184,64 @@ def build_elastic_search_query(
     query = {"size": count, "query": weighted_query.to_dict()}
 
     return query
+
+def build_elastic_more_like_this_query(
+    language: Language,
+    like_documents: list[dict[str, str]],
+    similar_to: list[str],
+    cutoff_age: int,
```
Comment (tfnribeiro) [inline zeeguu/core/elastic/elastic_query_builder.py:192]: `cutoff_days`, might be a better name for this one?

File: zeeguu/api/endpoints/user_articles.py [hunk 3]
```diff
@@ -94,3 +94,20 @@ def user_articles_cohort():
...
+# ---------------------------------------------------------------------------
+@cross_domain
+@with_session
+def user_articles_foryou():
+    article_infos = []
+    try:
+        articles = content_recommendations(flask.g.user.id, flask.g.user.learned_language_id)
+        print("Sending CB recommendations")
+    except:
+        #Usually no recommendations when the user has not liked any articles
```
Comment (tfnribeiro) [inline zeeguu/api/endpoints/user_articles.py:109]: We don't want to silently catch all errors, make sure to at least log the exception if there is an error.

Currently we use Sentry to keep track of exceptions in zeeguu:

```
    except Exception as e:
        from sentry_sdk import capture_exception

        capture_exception(e)
```

**Review by tfnribeiro — COMMENTED** [review]
Some extra comments

File: zeeguu/api/endpoints/user_articles.py (hunk 3 above)
Comment (nselpriv) [reply zeeguu/api/endpoints/user_articles.py:112]: It should be ready now

File: zeeguu/core/elastic/elastic_query_builder.py [hunk 4]
```diff
@@ -182,3 +184,64 @@ def build_elastic_search_query(
...
+    like_documents: list[dict[str, str]],
+    similar_to: list[str],
+    cutoff_days: int,
+    scale: str ="10d",
+    offset: str ="4h",
+    decay: float=0.9):
+    
+    cutoff_date = datetime.now() - timedelta(days=cutoff_days)
+
+    query = {
```
Comment (mircealungu) [inline zeeguu/core/elastic/elastic_query_builder.py:199]: json is not for humans !!! 
in other queries i built little python DSLs that help build up such a monster w/o having to dump the whole string in the code. i'll accept it like this for now, but if you find the time after you've defended, i'd be happy to receive an updated code version that is less a monster string and more a nicely constructed python expression.

**Review by mircealungu — APPROVED** [review]
only one comment, but i'll accept for now so we can merge and deploy! i want to start getting recommendations!

File: zeeguu/core/elastic/elastic_query_builder.py (hunk 4 above)
Comment (nselpriv) [reply zeeguu/core/elastic/elastic_query_builder.py:199]: I'll make sure to take a look at it again and extend the DSL with what I need and then build the same query with it.
Thanks for the feedback

**Pull request merged**